backend/dataall/base/api/graphql_api.py

```diff
@@ -0,0 +1,183 @@
+import inspect
+from dataclasses import dataclass, fields
+from enum import Enum
+from typing import get_type_hints
+
+from dataall.base.api import gql
+from dataall.base.utils.decorator_utls import process_func
+
+
+@dataclass
+class Page(gql.ObjectType):
+    count: int
+    page: int
+    pageSize: int
+    pages: int
+    hasNext: bool
+    hasPrevious: bool
+    previousPage: str
+    nextPage: str
+
+
+@dataclass
+class PageFilter(gql.InputType):
+    term: str = None
+    page: int = 1
+    pageSize: int = 15
+
+
+_scalars = (int, float, str, bool)
+
+
+_scalar_serialize_mapping = {
+    str: gql.String,
+    int: gql.Integer,
+    float: gql.Number,
+    bool: gql.Boolean,
+}
+
+_object_names = set()
+
+_objects = {}
+
+
+def _resolve_ref(ref_type):
+    ref_name = _objects.get(ref_type, None)
+    if not ref_name:
+        raise ValueError(f"No reference with name {ref_name}")
+
+
+def _resolve_type(arg_type):
+    if isinstance(arg_type, list):
+        return gql.ArrayType(_resolve_type(arg_type[0]))
+    if arg_type == gql.ID or isinstance(arg_type, gql.Ref):
+        return arg_type
+    if issubclass(arg_type, Enum):
+        return gql.Enum(name=arg_type.__name__, values=arg_type)
+    if issubclass(arg_type, gql.ObjectType) or issubclass(arg_type, gql.InputType):
+        return gql.Ref(_objects[arg_type])
+    return _scalar_serialize_mapping.get(arg_type)
+
+
+def _has_default_value(cls, field_name):
+    for field_info in fields(cls):
+        if field_info.name == field_name:
+            return field_info.default is not field_info.default_factory
+    return False
+
+
+def _create_custom_object(cls, data: dict):
+    # TODO: fix the problem of nested objects in the input
+    return cls(**data)
+
+
+def _process_api_func(name, field):
+    def decorator(f):
+        fn, fn_decorator = process_func(f)
+
+        hints = get_type_hints(fn)
+        signature = inspect.signature(fn)
+        num_args = len(signature.parameters)
+
+        if "return" not in hints:
+            raise ValueError(f"Function {fn.__name__} doesn't specify the return type")
+
```

Comment:
Ahh I see what you are doing. You are using hints to get the parameters that include hints and then we compare this with the signature parameters to check if everything contains a type hint! + check that return also includes hints

Reply:
Yes. Since GraphQL schema has to have type and I wanted to utilize this fact in the code. 
All functions that defines the schema should be typed and types must be scalars (bools, strs, ints, floats) or custom object.